src/recipe_dictionary.cpp

```diff
@@ -1,95 +1,289 @@
 #include "recipe_dictionary.h"
-#include "crafting.h"
-#include "requirements.h"
 
-#include <algorithm> //std::remove
+#include "itype.h"
+#include "generic_factory.h"
 
-using itype_id = std::string; // From itype.h
+#include <algorithm>
+#include <numeric>
 
 recipe_dictionary recipe_dict;
 
-void recipe_dictionary::finalize()
-{
-    delete_if( [&]( recipe & r ) {
-        return r.requirements().is_blacklisted();
-    } );
+static recipe null_recipe;
+static std::set<const recipe *> null_match;
 
-    for( auto r : recipes ) {
-        add_to_component_lookup( r );
-    }
+const recipe &recipe_dictionary::operator[]( const std::string &id ) const
+{
+    auto iter = recipes.find( id );
+    return iter != recipes.end() ? iter->second : null_recipe;
 }
 
-void recipe_dictionary::add( recipe *rec )
+const std::set<const recipe *> &recipe_dictionary::in_category( const std::string &cat ) const
 {
-    recipes.push_back( rec );
-    by_name[rec->ident()] = rec;
-    by_category[rec->cat].push_back( rec );
+    auto iter = category.find( cat );
+    return iter != category.end() ? iter->second : null_match;
 }
 
-void recipe_dictionary::remove( recipe *rec )
+const std::set<const recipe *> &recipe_dictionary::of_component( const itype_id &id ) const
 {
-    recipes.remove( rec );
-    remove_from_component_lookup( rec );
-    by_name.erase( rec->ident() );
-    // Terse name for category vector since it's repeated so many times.
-    auto &cat_vec = by_category[rec->cat];
-    cat_vec.erase( std::remove( cat_vec.begin(), cat_vec.end(), rec ), cat_vec.end() );
+    auto iter = component.find( id );
+    return iter != component.end() ? iter->second : null_match;
 }
 
-void recipe_dictionary::delete_if( const std::function<bool( recipe & )> &pred )
+void recipe_dictionary::load( JsonObject &jo, const std::string & /* src */, bool uncraft )
 {
-    for( auto iter = recipes.begin(); iter != recipes.end(); ) {
-        recipe *const r = *iter;
-        // Already moving to the next, so we can erase the recipe without invalidating `iter`.
-        ++iter;
-        if( pred( *r ) ) {
-            remove( r );
-            delete r;
+    // @todo enable strict parsing for core recipes
+    bool strict = false;
+
+    // @todo remove once recipes are converted
+    if( !uncraft ) {
+        uncraft = jo.get_string( "category" ) == "CC_NONCRAFT";
+    }
+
+    auto result = jo.get_string( "result" );
+    auto id = result + ( uncraft ? "uncraft" : jo.get_string( "id_suffix", "" ) );
+
+    auto &r = recipe_dict.recipes[ id ];
+    r.result = result;
+    r.ident_ = id;
+
+    if( uncraft ) {
+        r.cat = "CC_NONCRAFT";
+        r.subcat = "CSC_NONCRAFT";
+        r.reversible = true;
+        r.valid_to_learn = false;
```

Comment:
Some kind of `recipe::load()` would be useful here. All these `r.whatever` clearly signals that they want to be in the `recipe` class. They would also acquire some intimacy there (being private if needed).